File: kelmah-mobile-ios/Kelmah/Features/Jobs/Data/JobsRepository.swift
import Foundation

final class JobsRepository {
    private let apiClient: APIClient

    init(apiClient: APIClient) {
        self.apiClient = apiClient
    }

    func getJobs(filters: JobFilters, page: Int = 1, limit: Int = 20) async throws -> JobsPage {
        let queryItems = buildQueryItems(filters: filters, page: page, limit: limit)
        let response = try await apiClient.send(
            path: "jobs",
            method: .get,
            queryItems: queryItems,
            requiresAuth: true,
            responseType: JobsRawEnvelope.self
        )
        return parseJobsPage(response, forcedSaved: false)
    }

    func getRecommendedJobs(limit: Int = 6) async throws -> RecommendationFeed {
        let response = try await apiClient.send(
            path: "jobs/recommendations/personalized",
            method: .get,
            queryItems: [
                URLQueryItem(name: "page", value: "1"),
                URLQueryItem(name: "limit", value: String(limit)),
            ],
            requiresAuth: true,
            responseType: JobsRawEnvelope.self
        )
        let personalizedPage = parseJobsPage(response, forcedSaved: false, strictJobQuality: true)
        let recommendationSource = response.meta?.objectValue?["recommendationSource"]?.stringValue
            ?? response.data?.objectValue?["recommendationSource"]?.stringValue
        let isNewUser = response.data?.objectValue?["isNewUser"]?.boolValue ?? false
        let isProfileIncomplete = recommendationSource == "profile-incomplete" || (isNewUser && personalizedPage.jobs.isEmpty)

        if isProfileIncomplete {
            let generalRecommendations = await fetchGeneralRecommendations(limit: limit)
            return RecommendationFeed(
                jobs: generalRecommendations,
                state: .profileIncomplete,
                contextMessage: buildProfileIncompleteMessage(
                    serverMessage: response.message,
                    hasGeneralRecommendations: generalRecommendations.isEmpty == false
                )
            )
        }

        return RecommendationFeed(
            jobs: personalizedPage.jobs,
            state: .personalized,
            contextMessage: nil
        )
    }

    func getMyJobs(limit: Int = 6) async throws -> [JobSummary] {
        let response = try await apiClient.send(
            path: "jobs/my-jobs",
            method: .get,
            queryItems: [
                URLQueryItem(name: "page", value: "1"),
                URLQueryItem(name: "limit", value: String(limit)),
                URLQueryItem(name: "sort", value: "-updatedAt"),
            ],
            requiresAuth: true,
            responseType: JobsRawEnvelope.self
        )
        return parseJobsPage(response, forcedSaved: false).jobs
    }

    func getSavedJobs(page: Int = 1, limit: Int = 20) async throws -> JobsPage {
        let response = try await apiClient.send(
            path: "jobs/saved",
            method: .get,
            queryItems: [
                URLQueryItem(name: "page", value: String(page)),
                URLQueryItem(name: "limit", value: String(limit)),
            ],
            requiresAuth: true,
            responseType: JobsRawEnvelope.self
        )
        return parseJobsPage(response, forcedSaved: true)
    }

    func getCategories() async throws -> [JobCategory] {
        let response = try await apiClient.send(
            path: "jobs/categories",
            method: .get,
            requiresAuth: true,
            responseType: JobsRawEnvelope.self
        )
        return parseCategories(response)
    }

    func getJobDetail(jobId: String) async throws -> JobDetail {
        let response = try await apiClient.send(
            path: "jobs/\(jobId)",
            method: .get,
            requiresAuth: true,
            responseType: JobsRawEnvelope.self
        )
        return try parseJobDetail(response)
    }

    func toggleSaved(jobId: String, shouldSave: Bool) async throws -> Bool {
        let method: HTTPMethod = shouldSave ? .post : .delete
        _ = try await apiClient.send(
            path: "jobs/\(jobId)/save",
            method: method,
            requiresAuth: true,
            responseType: JobsRawEnvelope.self
        ) as JobsRawEnvelope
        return shouldSave
    }

    func applyToJob(jobId: String, request: ApplyToJobRequest) async throws -> JobApplicationResult {
        let response = try await apiClient.send(
            path: "jobs/\(jobId)/apply",
            method: .post,
            body: request,
            requiresAuth: true,
            responseType: JobsRawEnvelope.self
        )

        return JobApplicationResult(
            success: response.success ?? true,
            message: response.message ?? response.data?["message"]?.stringValue ?? "Application submitted successfully"
        )
    }

    private func buildQueryItems(filters: JobFilters, page: Int, limit: Int) -> [URLQueryItem] {
        var items = [
            URLQueryItem(name: "page", value: String(page)),
            URLQueryItem(name: "limit", value: String(limit)),
            URLQueryItem(name: "sort", value: filters.sort.rawValue),
        ]
        if filters.search.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false {
            items.append(URLQueryItem(name: "search", value: filters.search.trimmingCharacters(in: .whitespacesAndNewlines)))
        }
        if filters.category.isEmpty == false, filters.category != "All" {
            items.append(URLQueryItem(name: "category", value: filters.category))
        }
        if filters.location.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false {
            items.append(URLQueryItem(name: "location", value: filters.location.trimmingCharacters(in: .whitespacesAndNewlines)))
        }
        return items
    }

    private func parseJobsPage(_ response: JobsRawEnvelope, forcedSaved: Bool, strictJobQuality: Bool = false) -> JobsPage {
        let dataArray = response.data?.arrayValue
        let dataObject = response.data?.objectValue
        let items = dataArray
            ?? dataObject?["items"]?.arrayValue
            ?? dataObject?["jobs"]?.arrayValue
            ?? []

        let pagination = dataObject?["pagination"]?.objectValue
        let metaPagination = response.meta?.objectValue?["pagination"]?.objectValue

        return JobsPage(
            jobs: items.compactMap { parseJobSummary($0, forcedSaved: forcedSaved, strictJobQuality: strictJobQuality) },
            page: pagination?["page"]?.intValue ?? metaPagination?["page"]?.intValue ?? 1,
            totalPages: pagination?["totalPages"]?.intValue ?? metaPagination?["totalPages"]?.intValue ?? 1,
            totalItems: pagination?["total"]?.intValue ?? metaPagination?["total"]?.intValue ?? items.count
        )
    }

    private func parseCategories(_ response: JobsRawEnvelope) -> [JobCategory] {
        let values = response.data?.arrayValue ?? response.data?["items"]?.arrayValue ?? []
        return values.enumerated().compactMap { index, value in
            guard let object = value.objectValue, let name = object.string("name") else { return nil }
            return JobCategory(
                id: object.string("_id") ?? object.string("id") ?? "category-\(index)",
                name: name,
                description: object.string("description") ?? ""
            )
        }
    }

    private func parseJobDetail(_ response: JobsRawEnvelope) throws -> JobDetail {
        guard let object = response.data?.objectValue ?? response.data?.arrayValue?.first?.objectValue else {
            throw APIClientError.invalidStatusCode(500, "Job detail payload was invalid")
        }
        guard let summary = parseJobSummary(.object(object), forcedSaved: object.bool("isSaved") ?? false, strictJobQuality: false) else {
            throw APIClientError.invalidStatusCode(500, "Job detail payload was invalid")
        }

        let requirementsContainer = object["requirements"]?.objectValue
        let requirements = ((requirementsContainer?["primarySkills"]?.arrayValue ?? []) + (requirementsContainer?["secondarySkills"]?.arrayValue ?? []))
            .compactMap { $0.stringValue }
        return JobDetail(
            summary: summary,
            fullDescription: object.string("description") ?? summary.description,
            requirements: requirements.isEmpty ? summary.skills : requirements,
            proposalCount: object.int("proposalCount") ?? 0,
            viewCount: object.int("viewCount") ?? 0,
            deadline: object.string("deadline") ?? object.string("endDate"),
            hirerId: object["hirer"]?.objectValue?.string("_id") ?? object["hirer"]?.objectValue?.string("id") ?? object.string("hirerId")
        )
    }

    private func parseJobSummary(_ value: JSONValue, forcedSaved: Bool, strictJobQuality: Bool = false) -> JobSummary? {
        guard let object = value.objectValue else { return nil }
        guard let id = object.string("_id") ?? object.string("id") else { return nil }
        let title = object.string("title")?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let description = object.string("description")?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""

        let budgetObject = object["budget"]?.objectValue
        let employerObject = object["hirer"]?.objectValue
        let employerName = employerObject?.string("name")
            ?? [employerObject?.string("firstName"), employerObject?.string("lastName")]
                .compactMap { $0 }
                .joined(separator: " ")
                .trimmingCharacters(in: .whitespacesAndNewlines)
                .nilIfEmpty
            ?? object.string("hirer_name")
            ?? object.string("company")

        if strictJobQuality && (title.isEmpty || description.isEmpty || employerName?.isEmpty != false) {
            return nil
        }

        let paymentType = object.string("paymentType") ?? budgetObject?.string("type") ?? "fixed"
        let budgetAmount = object.double("budget") ?? budgetObject?.double("amount") ?? budgetObject?.double("max") ?? 0
        let currency = object.string("currency") ?? budgetObject?.string("currency") ?? "GHS"

        return JobSummary(
            id: id,
            title: title.isEmpty ? "Untitled Job" : title,
            description: description,
            category: object.string("category") ?? "General",
            locationLabel: locationLabel(from: object),
            budgetLabel: formatBudgetLabel(amount: budgetAmount, currency: currency, paymentType: paymentType),
            budgetAmount: budgetAmount,
            currency: currency,
            paymentType: paymentType,
            employerName: employerName ?? "Employer Name Pending",
            employerAvatar: employerObject?.string("avatar") ?? employerObject?.string("profileImage"),
            skills: skills(from: object),
            postedAt: object.string("createdAt") ?? object.string("created_at") ?? object.string("postedDate"),
            status: object.string("status"),
            proposalCount: object.int("proposalCount") ?? object.int("applicationsCount") ?? 0,
            matchScore: object.double("matchScore") ?? object.int("matchScore").map(Double.init),
            aiReasoning: parseAIReasoning(from: object),
            isVerified: employerObject?.bool("verified") ?? employerObject?.bool("isVerified") ?? false,
            isUrgent: object.bool("urgent") ?? false,
            isSaved: forcedSaved || (object.bool("isSaved") ?? object.bool("saved") ?? false)
        )
    }

    private func locationLabel(from object: [String: JSONValue]) -> String {
        if let inlineLocation = object.string("location")?.nilIfEmpty {
            return inlineLocation
        }
        let location = object["location"]?.objectValue
        let locationDetails = object["locationDetails"]?.objectValue
        let labels = [
            location?.string("city"),
            location?.string("region") ?? locationDetails?.string("region"),
            location?.string("country")
        ].compactMap { $0?.nilIfEmpty }
        if labels.isEmpty == false {
            return labels.joined(separator: ", ")
        }
        return location?.string("address")
            ?? locationDetails?.string("district")
            ?? location?.string("type")?.capitalized
            ?? "Location not specified"
    }

    private func skills(from object: [String: JSONValue]) -> [String] {
        (object["skills"]?.arrayValue ?? []).compactMap { skill in
            if let text = skill.stringValue { return text.nilIfEmpty }
            if let value = skill.objectValue?.string("name") ?? skill.objectValue?.string("label") ?? skill.objectValue?.string("type") {
                return value.nilIfEmpty
            }
            return nil
        }
    }

    private func formatBudgetLabel(amount: Double, currency: String, paymentType: String) -> String {
        let numberText: String
        if amount.truncatingRemainder(dividingBy: 1) == 0 {
            numberText = String(Int(amount))
        } else {
            numberText = String(format: "%.2f", amount)
        }
        let suffix = paymentType.lowercased() == "hourly" ? "/hr" : ""
        return "\(currency) \(numberText)\(suffix)"
    }

    private func parseAIReasoning(from object: [String: JSONValue]) -> String? {
        if let inlineReasoning = object.string("aiReasoning")?.trimmingCharacters(in: .whitespacesAndNewlines), inlineReasoning.isEmpty == false {
            return inlineReasoning
        }

        return object["aiReasons"]?.arrayValue?
            .compactMap { $0.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines) }
            .first(where: { $0.isEmpty == false })
    }

    private func fetchGeneralRecommendations(limit: Int) async -> [JobSummary] {
        do {
            let response = try await apiClient.send(
                path: "jobs/recommendations",
                method: .get,
                queryItems: [
                    URLQueryItem(name: "page", value: "1"),
                    URLQueryItem(name: "limit", value: String(limit)),
                ],
                requiresAuth: true,
                responseType: JobsRawEnvelope.self
            )
            return parseJobsPage(response, forcedSaved: false, strictJobQuality: true).jobs
        } catch {
            return []
        }
    }

    private func buildProfileIncompleteMessage(serverMessage: String?, hasGeneralRecommendations: Bool) -> String {
        let baseMessage = serverMessage?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
            ?? "Complete your profile to unlock personalized job matches."
        if hasGeneralRecommendations {
            return "\(baseMessage) Showing general recommendations while you complete your profile."
        }
        return baseMessage
    }
}

private extension Dictionary where Key == String, Value == JSONValue {
    func string(_ key: String) -> String? { self[key]?.stringValue }
    func int(_ key: String) -> Int? { self[key]?.intValue }
    func double(_ key: String) -> Double? { self[key]?.doubleValue }
    func bool(_ key: String) -> Bool? { self[key]?.boolValue }
}

private extension String {
    var nilIfEmpty: String? {
        trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : self
    }
}

File: kelmah-mobile-android/app/src/main/java/com/kelmah/mobile/features/jobs/data/JobsRepository.kt
package com.kelmah.mobile.features.jobs.data

import com.kelmah.mobile.core.network.ApiResult
import com.kelmah.mobile.core.session.SessionCoordinator
import dagger.Lazy
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import retrofit2.HttpException

@Singleton
class JobsRepository @Inject constructor(
    private val jobsApiService: JobsApiService,
    private val sessionCoordinator: Lazy<SessionCoordinator>,
) {

    suspend fun getJobs(
        filters: JobsFilterState,
        page: Int = 1,
        limit: Int = 20,
    ): ApiResult<JobsPage> = executeAuthorized {
        val response = jobsApiService.getJobs(
            buildMap {
                put("page", page.toString())
                put("limit", limit.toString())
                put("sort", filters.sort.queryValue)
                if (filters.search.isNotBlank()) put("search", filters.search.trim())
                if (filters.category.isNotBlank() && filters.category != "All") put("category", filters.category)
                if (filters.location.isNotBlank()) put("location", filters.location.trim())
            },
        )
        ApiResult.Success(parseJobsPage(response))
    }

    suspend fun getRecommendedJobs(limit: Int = 6): ApiResult<RecommendationFeed> = executeAuthorized {
        val response = jobsApiService.getRecommendedJobs(
            mapOf(
                "page" to "1",
                "limit" to limit.toString(),
            ),
        )
        val personalizedJobs = parseJobsPage(response, strictJobQuality = true).jobs
        val recommendationSource = response.nestedObject("meta")?.string("recommendationSource")
        val isProfileIncomplete = recommendationSource.equals("profile-incomplete", ignoreCase = true) ||
            (response.nestedObject("data")?.bool("isNewUser") == true && personalizedJobs.isEmpty())

        if (isProfileIncomplete) {
            val generalRecommendations = fetchGeneralRecommendations(limit)
            ApiResult.Success(
                RecommendationFeed(
                    jobs = generalRecommendations,
                    state = RecommendationFeedState.PROFILE_INCOMPLETE,
                    contextMessage = buildProfileIncompleteMessage(
                        serverMessage = response.string("message"),
                        hasGeneralRecommendations = generalRecommendations.isNotEmpty(),
                    ),
                ),
            )
        } else {
            ApiResult.Success(
                RecommendationFeed(
                    jobs = personalizedJobs,
                    state = RecommendationFeedState.PERSONALIZED,
                ),
            )
        }
    }

    suspend fun getMyJobs(limit: Int = 6): ApiResult<List<JobSummary>> = executeAuthorized {
        val response = jobsApiService.getMyJobs(
            mapOf(
                "page" to "1",
                "limit" to limit.toString(),
                "sort" to "-updatedAt",
            ),
        )
        ApiResult.Success(parseJobsPage(response).jobs)
    }

    suspend fun getSavedJobs(
        page: Int = 1,
        limit: Int = 20,
    ): ApiResult<JobsPage> = executeAuthorized {
        val response = jobsApiService.getSavedJobs(
            mapOf(
                "page" to page.toString(),
                "limit" to limit.toString(),
            ),
        )
        ApiResult.Success(parseJobsPage(response, forcedSaved = true))
    }

    suspend fun getCategories(): ApiResult<List<JobCategory>> = executeAuthorized {
        val response = jobsApiService.getCategories()
        ApiResult.Success(parseCategories(response))
    }

    suspend fun getJobDetail(jobId: String): ApiResult<JobDetail> = executeAuthorized {
        val response = jobsApiService.getJobById(jobId)
        ApiResult.Success(parseJobDetail(response))
    }

    suspend fun toggleSaved(jobId: String, shouldSave: Boolean): ApiResult<Boolean> = executeAuthorized {
        if (shouldSave) {
            jobsApiService.saveJob(jobId)
        } else {
            jobsApiService.unsaveJob(jobId)
        }
        ApiResult.Success(shouldSave)
    }

    suspend fun applyToJob(jobId: String, request: ApplyToJobRequest): ApiResult<JobApplicationResult> = executeAuthorized {
        val response = jobsApiService.applyToJob(jobId, request)
        ApiResult.Success(
            JobApplicationResult(
                success = response.string("success")?.toBooleanStrictOrNull() ?: true,
                message = response.string("message") ?: response.nestedObject("data")?.string("message") ?: "Application submitted successfully",
            ),
        )
    }

    private suspend fun <T> executeAuthorized(block: suspend () -> ApiResult<T>): ApiResult<T> {
        return try {
            block()
        } catch (error: HttpException) {
            if (error.code() == 401 && sessionCoordinator.get().refreshSession()) {
                try {
                    block()
                } catch (retryError: Exception) {
                    ApiResult.Error(message = retryError.message ?: "Request failed after session refresh")
                }
            } else {
                ApiResult.Error(message = error.message ?: "Request failed", code = error.code())
            }
        } catch (error: Exception) {
            ApiResult.Error(message = error.message ?: "Request failed")
        }
    }

    private fun parseJobsPage(
        response: JsonObject,
        forcedSaved: Boolean = false,
        strictJobQuality: Boolean = false,
    ): JobsPage {
        val dataNode = response["data"]
        val dataObject = dataNode as? JsonObject
        val items = when (dataNode) {
            is JsonArray -> dataNode
            is JsonObject -> dataNode.nestedArray("items")
                ?: dataNode.nestedArray("jobs")
                ?: JsonArray(emptyList())
            else -> response.nestedArray("items")
                ?: response.nestedArray("jobs")
                ?: JsonArray(emptyList())
        }

        val pagination = dataObject?.nestedObject("pagination")
        val metaPagination = response.nestedObject("meta")?.nestedObject("pagination")
        val page = pagination?.int("page")
            ?: metaPagination?.int("page")
            ?: response.int("page")
            ?: 1
        val totalPages = pagination?.int("totalPages")
            ?: metaPagination?.int("totalPages")
            ?: response.int("totalPages")
            ?: 1
        val totalItems = pagination?.int("total")
            ?: metaPagination?.int("total")
            ?: response.int("total")
            ?: items.size

        return JobsPage(
            jobs = items.mapNotNull { parseJobSummary(it, forcedSaved, strictJobQuality) },
            page = page,
            totalPages = totalPages,
            totalItems = totalItems,
        )
    }

    private fun parseCategories(response: JsonObject): List<JobCategory> {
        val dataNode = response["data"]
        val items = when (dataNode) {
            is JsonArray -> dataNode
            is JsonObject -> dataNode["items"] as? JsonArray ?: JsonArray(emptyList())
            else -> JsonArray(emptyList())
        }

        return items.mapIndexedNotNull { index, item ->
            val obj = item as? JsonObject ?: return@mapIndexedNotNull null
            val name = obj.string("name") ?: return@mapIndexedNotNull null
            JobCategory(
                id = obj.string("_id") ?: obj.string("id") ?: "category-$index",
                name = name,
                description = obj.string("description") ?: "",
            )
        }
    }

    private fun parseJobDetail(response: JsonObject): JobDetail {
        val raw = response.nestedObject("data") ?: response
        val summary = parseJobSummary(raw, forcedSaved = raw.bool("isSaved") ?: false, strictJobQuality = false)
            ?: throw IllegalStateException("Job detail payload was invalid")

        val fullDescription = raw.string("description") ?: summary.description
        val requirements = buildList {
            val primarySkills = raw.nestedObject("requirements")?.nestedArray("primarySkills")
                ?.mapNotNull { it.primitiveContentOrNull() }
                ?: emptyList()
            val secondarySkills = raw.nestedObject("requirements")?.nestedArray("secondarySkills")
                ?.mapNotNull { it.primitiveContentOrNull() }
                ?: emptyList()
            addAll(primarySkills)
            addAll(secondarySkills)
        }.distinct()

        return JobDetail(
            summary = summary,
            fullDescription = fullDescription,
            requirements = requirements.ifEmpty { summary.skills },
            proposalCount = raw.int("proposalCount") ?: 0,
            viewCount = raw.int("viewCount") ?: 0,
            deadline = raw.string("deadline") ?: raw.string("endDate"),
            hirerId = raw.nestedObject("hirer")?.string("_id")
                ?: raw.nestedObject("hirer")?.string("id")
                ?: raw.string("hirerId"),
        )
    }

    private fun parseJobSummary(
        element: JsonElement,
        forcedSaved: Boolean = false,
        strictJobQuality: Boolean = false,
    ): JobSummary? {
        val job = element as? JsonObject ?: return null
        val id = job.string("_id") ?: job.string("id") ?: return null
        val title = job.string("title")?.trim().orEmpty()
        val description = (job.string("description") ?: "").trim()
        val budgetValue = job.double("budget")
            ?: job.nestedObject("budget")?.double("amount")
            ?: job.nestedObject("budget")?.double("max")
            ?: 0.0
        val currency = job.string("currency")
            ?: job.nestedObject("budget")?.string("currency")
            ?: "GHS"
        val employer = job.nestedObject("hirer")
        val employerName = employer?.string("name")
            ?: job.string("hirer_name")
            ?: listOfNotNull(employer?.string("firstName"), employer?.string("lastName")).joinToString(" ").ifBlank { null }
            ?: job.string("company")

        if (strictJobQuality && (title.isBlank() || description.isBlank() || employerName.isNullOrBlank())) {
            return null
        }

        return JobSummary(
            id = id,
            title = title.ifBlank { "Untitled Job" },
            description = description,
            category = job.string("category") ?: "General",
            locationLabel = parseLocationLabel(job),
            budgetLabel = formatBudgetLabel(
                amount = budgetValue,
                currency = currency,
                paymentType = job.string("paymentType") ?: job.nestedObject("budget")?.string("type") ?: "fixed",
            ),
            budgetAmount = budgetValue,
            currency = currency,
            paymentType = job.string("paymentType") ?: job.nestedObject("budget")?.string("type") ?: "fixed",
            employerName = employerName ?: "Employer Name Pending",
            employerAvatar = employer?.string("avatar") ?: employer?.string("profileImage"),
            skills = parseSkills(job),
            postedAt = job.string("createdAt") ?: job.string("created_at") ?: job.string("postedDate"),
            status = job.string("status"),
            proposalCount = job.int("proposalCount") ?: job.int("applicationsCount") ?: 0,
            matchScore = job.double("matchScore") ?: job.int("matchScore")?.toDouble(),
            aiReasoning = parseAiReasoning(job),
            isVerified = employer?.bool("verified") ?: employer?.bool("isVerified") ?: false,
            isUrgent = job.bool("urgent") ?: false,
            isSaved = forcedSaved || job.bool("isSaved") ?: job.bool("saved") ?: false,
        )
    }

    private fun parseLocationLabel(job: JsonObject): String {
        job.string("location")?.takeIf { it.isNotBlank() }?.let { return it }
        val location = job["location"] as? JsonObject
        val locationDetails = job["locationDetails"] as? JsonObject
        val composed = listOfNotNull(
            location?.string("city"),
            location?.string("region") ?: locationDetails?.string("region"),
            location?.string("country"),
        ).filter { it.isNotBlank() }
        if (composed.isNotEmpty()) return composed.joinToString(", ")
        return location?.string("address")
            ?: locationDetails?.string("district")
            ?: location?.string("type")?.replaceFirstChar { it.uppercase() }
            ?: "Location not specified"
    }

    private fun parseSkills(job: JsonObject): List<String> {
        return job.nestedArray("skills")
            ?.mapNotNull { skill ->
                when (skill) {
                    is JsonPrimitive -> skill.contentOrNull()
                    is JsonObject -> skill.string("name") ?: skill.string("label") ?: skill.string("type")
                    else -> null
                }
            }
            ?.filter { it.isNotBlank() }
            ?: emptyList()
    }

    private fun formatBudgetLabel(amount: Double, currency: String, paymentType: String): String {
        val normalizedAmount = if (amount % 1.0 == 0.0) amount.toInt().toString() else String.format("%.2f", amount)
        val suffix = if (paymentType.equals("hourly", ignoreCase = true)) "/hr" else ""
        return "$currency $normalizedAmount$suffix"
    }

    private fun parseAiReasoning(job: JsonObject): String? {
        val inlineReasoning = job.string("aiReasoning")?.trim()
        if (inlineReasoning.isNullOrBlank().not()) {
            return inlineReasoning
        }

        val reasons = job.nestedArray("aiReasons")
            ?.mapNotNull { it.primitiveContentOrNull()?.trim() }
            ?.filter { it.isNotBlank() }
            .orEmpty()

        return reasons.firstOrNull()
    }

    private suspend fun fetchGeneralRecommendations(limit: Int): List<JobSummary> {
        return try {
            val response = jobsApiService.getGeneralRecommendations(
                mapOf(
                    "page" to "1",
                    "limit" to limit.toString(),
                ),
            )
            parseJobsPage(response, strictJobQuality = true).jobs
        } catch (_: Exception) {
            emptyList()
        }
    }

    private fun buildProfileIncompleteMessage(serverMessage: String?, hasGeneralRecommendations: Boolean): String {
        val baseMessage = serverMessage?.takeIf { it.isNotBlank() }
            ?: "Complete your profile to unlock personalized job matches."
        return if (hasGeneralRecommendations) {
            "$baseMessage Showing general recommendations while you complete your profile."
        } else {
            baseMessage
        }
    }
}

private fun JsonObject.string(key: String): String? = (this[key] as? JsonPrimitive)?.contentOrNull()

private fun JsonObject.int(key: String): Int? = (this[key] as? JsonPrimitive)?.intOrNull()

private fun JsonObject.double(key: String): Double? = (this[key] as? JsonPrimitive)?.doubleOrNull()

private fun JsonObject.bool(key: String): Boolean? = (this[key] as? JsonPrimitive)?.booleanOrNull()

private fun JsonObject.nestedObject(key: String): JsonObject? = this[key] as? JsonObject

private fun JsonObject.nestedArray(key: String): JsonArray? = this[key] as? JsonArray

private fun JsonPrimitive.contentOrNull(): String? = if (this == JsonNull) null else content

private fun JsonPrimitive.intOrNull(): Int? = content.toIntOrNull()

private fun JsonPrimitive.doubleOrNull(): Double? = content.toDoubleOrNull()

private fun JsonPrimitive.booleanOrNull(): Boolean? = content.toBooleanStrictOrNull()

private fun JsonElement.primitiveContentOrNull(): String? = (this as? JsonPrimitive)?.contentOrNull()
